File: 二叉树/把二叉搜索树转换为累加树.go
package 二叉树

//反中序遍历
func convertBST(root *TreeNode) *TreeNode {
	var convertInorder func(root *TreeNode)
	curSum := 0 //全局变量记录当前遍历所有节点的累加和
	convertInorder = func(root *TreeNode) {
		if root == nil {
			return
		}
		convertInorder(root.Right)
		tmp := root.Val //先存一下当前节点的初始值
		root.Val += curSum
		curSum += tmp
		convertInorder(root.Left)
	}
	convertInorder(root)
	return root
}

File: 二叉树/把二叉搜索树转换为累加树.java
package 二叉树;

//剑指offer
public class 把二叉搜索树转换为累加树 {
    int sum = 0;

    public TreeNode convertBST(TreeNode root) {
        dfs(root);
        return root;
    }

    private void dfs(TreeNode root) {
        if (root == null) return;
        dfs(root.right);
        int tmp = root.val;
        root.val += sum;
        sum += tmp;
        dfs(root.left);
    }
}
